src/main/java/fi/helsinki/cs/tmc/core/communication/TmcServerCommunicationTaskFactory.java

```diff
@@ -229,6 +236,50 @@ private URI getUnlockUrl(Course course) throws NotLoggedInException {
         return new HttpTasks().getForBinary(zipUrl);
     }
 
+    public Callable<SubmissionResponse> getSubmittingExerciseToSkillifierTask(
+            final Exercise exercise, final byte[] byteToSend, Map<String, String> extraParams) {
+
+        final Map<String, String> params = new LinkedHashMap<>();
+        params.put("client_time", "" + (System.currentTimeMillis() / 1000L));
+        params.put("client_nanotime", "" + System.nanoTime());
+        params.putAll(extraParams);
+
+        return wrapWithNotLoggedInException(new Callable<SubmissionResponse>() {
+            @Override
+            public SubmissionResponse call() throws Exception {
+                String response;
+                try {
+                    URI submitUrl = URI.create("http://localhost:3200/submit");
+                    //final URI submitUrl = addApiCallQueryParameters(exercise.getReturnUrl());
+                    final Callable<String> upload = new HttpTasks()
+                            .uploadRawDataForTextDownload(submitUrl, params, byteToSend);
+                    response = upload.call();
+                } catch (FailedHttpResponseException ex) {
+                    return checkForObsoleteClient(ex);
+                }
+
+                JsonObject respJson = new JsonParser().parse(response).getAsJsonObject();
+                if (respJson.get("error") != null) {
+                    throw new RuntimeException(
+                        "Server responded with error: " + respJson.get("error"));
+                } else if (respJson.get("submission_url") != null) {
+                    try {
+                        URI submissionUrl = new URI(respJson.get("submission_url").getAsString());
+                        URI pasteUrl = new URI(respJson.get("paste_url").getAsString());
+                        return new SubmissionResponse(submissionUrl, pasteUrl);
+                    } catch (Exception e) {
+                        throw new RuntimeException(
+                            "Server responded with malformed " + "submission url");
```

Comment:
Is this truly the one possible reason for an exception to occur here?

Also, wrap the exception (that is, retain the stack trace and the original message) rather than discarding and rethrowing. Consider logging the malformed response as well.